### PR TITLE
trace: add new command line options for enabling tracing at startup.

### PR DESCRIPTION
**Pull request by pwhelan**
Add new command line options to enable tracing from the start:
    
* --trace: define a complete trace pipeline in a single line.
* --trace-input: define the input to trace.
* --trace-output: define the trace output.
* --trace-output-property: define a trace output property.

The format for the `--trace` option is  the following:

```input=$INPUT_NAME output=$OUTPUT_PLUGIN output.$PROPERTY=$VALUE```

Several properties can be defined for the output plugin. Support for spaces in property values is not supported yet...

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [ ] Debug log output from testing the change

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [ ] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: src/fluent-bit.c [hunk 1]
```diff
@@ -718,6 +733,164 @@ static struct flb_cf *service_configure(struct flb_cf *cf,
...
+    }
+
+    key = mk_string_copy_substr(kv, 0, sep);
+    value = kv + sep + 1;
+
+    if (!key) {
+        return -1;
+    }
+
+    flb_kv_item_create_len(props,
```
Comment (Syn3rman) [inline src/fluent-bit.c:807]: I don't see a corresponding destroy for this, are we sure this is being cleaned up properly?

**Review by Syn3rman — CHANGES_REQUESTED** [review]
I've added a few comments and requested a few changes. We are ignoring the return values from all the functions and while I have pointed out only one instance, we do need to check it everywhere.

**Comment by leonardo-albertovich** [comment]
@pwhelan please let me know once you have addressed these requests widely in the code.

I've noticed many places where the results are not being checked but it makes no sense to point anything out before you have a chance to go over your changes and fix it.

File: src/fluent-bit.c [hunk 2]
```diff
@@ -985,6 +1168,28 @@ int flb_main(int argc, char **argv)
...
+                flb_free(trace_output);
+            }
+            trace_output = flb_strdup(optarg);
+            break;
+        case FLB_LONG_TRACE_OUTPUT_PROPERTY:
+            if (trace_props == NULL) {
+                trace_props = flb_calloc(1, sizeof(struct mk_list));
+                flb_kv_init(trace_props);
+            }
+            set_trace_property(trace_props, optarg);
```
[the remark this answers is not in the document]
Comment (pwhelan) [reply src/fluent-bit.c:1191]: I've add error checking to all the functions now.

**Comment by pwhelan** [comment]
Valgrind log:

```bash
==611685== Memcheck, a memory error detector
==611685== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==611685== Using Valgrind-3.21.0 and LibVEX; rerun with -h for copyright info
==611685== Command: ./bin/fluent-bit -Z -i dummy -o exit -p flush_count=5 -f 1 --trace=input=dummy.0\ output=null
==611685== Parent PID: 563866
==611685== 
==611685== 
==611685== HEAP SUMMARY:
==611685==     in use at exit: 0 bytes in 0 blocks
==611685==   total heap usage: 3,280 allocs, 3,280 frees, 4,246,283 bytes allocated
==611685== 
==611685== All heap blocks were freed -- no leaks are possible
==611685== 
==611685== For lists of detected and suppressed errors, rerun with: -s
==611685== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by pwhelan** [comment]
Here is a valgrind log after the latest changes:

```
==31986== Memcheck, a memory error detector
==31986== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==31986== Using Valgrind-3.21.0 and LibVEX; rerun with -h for copyright info
==31986== Command: ./bin/fluent-bit -Z -i dummy -o stdout -m * -o exit -m * -p flush_count=5 -f 1 --trace=input=dummy.0\ output=null
==31986== Parent PID: 2603
==31986== 
==31986== 
==31986== HEAP SUMMARY:
==31986==     in use at exit: 0 bytes in 0 blocks
==31986==   total heap usage: 3,637 allocs, 3,637 frees, 4,916,409 bytes allocated
==31986== 
==31986== All heap blocks were freed -- no leaks are possible
==31986== 
==31986== For lists of detected and suppressed errors, rerun with: -s
==31986== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by edsiper** [comment]
@pwhelan do you have a PR for docs ?

**Comment by pwhelan** [comment]
> do you have a PR for docs ?

@edsiper not yet, open one ASAP.

**Pull request merged**